File: server-test/src/test/java/de/njsm/stocks/servertest/v2/FoodTest.java
/*
 * stocks is client-server program to manage a household's food stock
 * Copyright (C) 2019  The stocks developers
 *
 * This file is part of the stocks program suite.
 *
 * stocks is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * stocks is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 *
 */

package de.njsm.stocks.servertest.v2;

import de.njsm.stocks.client.business.entities.IdImpl;
import de.njsm.stocks.client.business.entities.Location;
import de.njsm.stocks.servertest.TestSuite;
import de.njsm.stocks.servertest.v2.repo.LocationRepository;
import io.restassured.http.ContentType;
import io.restassured.response.ValidatableResponse;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Order;
import org.junit.jupiter.api.Test;

import javax.inject.Inject;

import static io.restassured.RestAssured.given;
import static org.hamcrest.Matchers.*;

@Order(800)
public class FoodTest extends Base implements Deleter {

    private LocationRepository locationRepository;

    @BeforeEach
    void setUp() {
        dagger.inject(this);
    }

    @Test
    void testBitemporalFood() {
        addFoodType(getUniqueName("testBitemporalFood"));

        assertOnFood(true)
                .body("data.validTimeStart", not(emptyIterable()))
                .body("data.validTimeEnd", not(emptyIterable()))
                .body("data.transactionTimeStart", not(emptyIterable()))
                .body("data.transactionTimeEnd", not(emptyIterable()))
                .body("data.toBuy", hasItem(false));
    }

    @Test
    void addAnItem() {
        String name = getUniqueName("addAnItem");
        addFoodType(name);

        assertOnFood()
                .body("data.name", hasItem(name));
    }

    @Test
    void renameFood() {
        String name = getUniqueName("renameFood");
        String newName = name + ".new";
        int id = createNewFoodType(name);
        IdImpl<Location> locationId = locationRepository.createNewLocationType(name);

        assertOnRename(id, 0, newName, 42, locationId.id())
                .statusCode(200)
                .body("status", equalTo(0));

        assertOnFood()
                .body("data.name", hasItem(newName))
                .body("data.expirationOffset", hasItem(42))
                .body("data.location", hasItem(locationId.id()));
    }

    @Test
    void renameFoodWithDescription() {
        String name = getUniqueName("renameFoodWithDescription");
        String newName = name + ".new";
        int id = createNewFoodType(name);
        IdImpl<Location> locationId = locationRepository.createNewLocationType(name);
        String description = "description";

        assertOnEdit(id, 0, newName, 42, locationId.id(), description)
                .statusCode(200)
                .body("status", equalTo(0));

        assertOnFood()
                .body("data.name", hasItem(newName))
                .body("data.expirationOffset", hasItem(42))
                .body("data.description", hasItem(description))
                .body("data.location", hasItem(locationId.id()));
    }

    @Test
    void alterFoodDescription() {
        String name = getUniqueName("alterFoodDescription");
        String newDescription = "new description";
        int id = createNewFoodType(name);

        given()
                .log().ifValidationFails()
                .queryParam("id", id)
                .queryParam("version", 0)
                .formParam("description", newDescription).
        when()
                .post(TestSuite.DOMAIN + "/v2/food/description").
        then()
                .log().ifValidationFails()
                .contentType(ContentType.JSON)
                .statusCode(200)
                .body("status", equalTo(0));

        assertOnFood()
                .body("data.description", hasItem(newDescription));
    }

    @Test
    void setBuyStatus() {
        String name = getUniqueName("setBuyStatus");
        int id = createNewFoodType(name);

        assertOnSetBuyStatus(id, 0, true)
                .body("status", equalTo(0));

        assertOnFood()
                .body("data.name", hasItem(name))
                .body("data.toBuy", hasItem(true));

    }

    @Test
    void renamingFailsWithWrongVersion() {
        String name = getUniqueName("renamingFailsWithWrongVersion");
        String newName = name + ".new";
        int id = createNewFoodType(name);

        assertOnRename(id, 99, newName, 0, 0)
                .statusCode(400)
                .body("status", equalTo(3));
    }

    @Test
    void renamingUnknownIdIsReported() {
        assertOnRename(9999, 0, "dummy", 0, 0)
                .statusCode(404)
                .body("status", equalTo(2));
    }

    @Test
    void deleteFood() {
        String name = getUniqueName("deleteFood");
        int id = createNewFoodType(name);

        assertOnDelete(id, 0)
                .statusCode(200)
                .body("status", equalTo(0));
    }

    @Test
    void deletingFailsWithWrongVersion() {
        String name = getUniqueName("deletingFailsWithWrongVersion");
        int id = createNewFoodType(name);

        assertOnDelete(id, 99)
                .statusCode(400)
                .body("status", equalTo(3));
    }

    @Test
    void deletingUnknownIdIsReported() {
        assertOnDelete(9999, 0)
                .statusCode(404)
                .body("status", equalTo(2));
    }

    public static int createNewFoodType(String name) {
        addFoodType(name);
        return getIdOfFood(name);
    }

    private static void addFoodType(String name) {
        given()
                .log().ifValidationFails()
                .queryParam("name", name)
        .when()
                .put(TestSuite.DOMAIN + "/v2/food").
        then()
                .log().ifValidationFails()
                .statusCode(200)
                .contentType(ContentType.JSON)
                .body("status", equalTo(0));
    }

    private static int getIdOfFood(String name) {
        return assertOnFood()
                .extract()
                .jsonPath()
                .getInt("data.findAll{ it.name == '" + name + "' }.last().id");

    }

    private ValidatableResponse assertOnRename(int id, int version, String newName, int expirationOffset, int location) {
        return
        given()
                .log().ifValidationFails()
                .queryParam("id", id)
                .queryParam("version", version)
                .queryParam("new", newName)
                .queryParam("expirationoffset", expirationOffset)
                .queryParam("location", location).
        when()
                .put(TestSuite.DOMAIN + "/v2/food/rename").
        then()
                .log().ifValidationFails()
                .contentType(ContentType.JSON);
    }

    private ValidatableResponse assertOnEdit(int id, int version, String newName, int expirationOffset, int location, String description) {
        return
        given()
                .log().ifValidationFails()
                .queryParam("id", id)
                .queryParam("version", version)
                .queryParam("new", newName)
                .queryParam("expirationoffset", expirationOffset)
                .queryParam("location", location)
                .formParam("description", description).
        when()
                .put(TestSuite.DOMAIN + "/v2/food/edit").
        then()
                .log().ifValidationFails()
                .contentType(ContentType.JSON);
    }

    private ValidatableResponse assertOnSetBuyStatus(int id, int version, boolean toBuy) {
        return
        given()
                .log().ifValidationFails()
                .queryParam("id", id)
                .queryParam("version", version)
                .queryParam("buy", toBuy ? 1 : 0).
        when()
                .put(TestSuite.DOMAIN + "/v2/food/buy").
        then()
                .log().ifValidationFails()
                .statusCode(200)
                .contentType(ContentType.JSON);
    }

    private static ValidatableResponse assertOnFood() {
        return assertOnFood(false);
    }

    public static ValidatableResponse assertOnFood(boolean bitemporal) {
        return
        given()
                .log().ifValidationFails()
                .queryParam("bitemporal", bitemporal ? 1 : 0).
        when()
                .get(TestSuite.DOMAIN + "/v2/food").
        then()
                .log().ifValidationFails()
                .statusCode(200)
                .contentType(ContentType.JSON)
                .body("status", equalTo(0));
    }

    @Override
    public String getEndpoint() {
        return "/v2/food";
    }

    @Inject
    void setLocationRepository(LocationRepository locationRepository) {
        this.locationRepository = locationRepository;
    }
}

File: server-test/src/test/java/de/njsm/stocks/servertest/v2/FoodTest.kt
/*
 * stocks is client-server program to manage a household's food stock
 * Copyright (C) 2019  The stocks developers
 *
 * This file is part of the stocks program suite.
 *
 * stocks is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * stocks is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 *
 */
package de.njsm.stocks.servertest.v2

import de.njsm.stocks.client.business.Constants
import de.njsm.stocks.client.business.EntityDeleteService
import de.njsm.stocks.client.business.FoodAddService
import de.njsm.stocks.client.business.FoodEditService
import de.njsm.stocks.client.business.StatusCodeException
import de.njsm.stocks.client.business.entities.Food
import de.njsm.stocks.client.business.entities.FoodAddForm
import de.njsm.stocks.client.business.entities.FoodForDeletion
import de.njsm.stocks.client.business.entities.FoodForEditing
import de.njsm.stocks.client.business.entities.FoodForSynchronisation
import de.njsm.stocks.client.business.entities.StatusCode
import de.njsm.stocks.servertest.v2.repo.LocationRepository
import de.njsm.stocks.servertest.v2.repo.UnitRepository
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatExceptionOfType
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Order
import org.junit.jupiter.api.Test
import java.time.Instant
import java.time.Period
import java.util.Optional
import javax.inject.Inject

@Order(800)
class FoodTest : Base() {
    internal lateinit var foodAddService: FoodAddService
        @Inject set

    internal lateinit var foodEditService: FoodEditService
        @Inject set

    internal lateinit var foodDeleteService: EntityDeleteService<Food>
        @Inject set

    internal lateinit var locationRepository: LocationRepository
        @Inject set

    internal lateinit var unitRepository: UnitRepository
        @Inject set

    @BeforeEach
    fun setUp() {
        dagger.inject(this)
    }

    @Test
    fun addFood() {
        val input = FoodAddForm.create(uniqueName, false, Period.ofDays(0), null, unitRepository.anyUnitId.id(), "")

        val id = foodAddService.add(input)

        val foods = updateService.getFood(Instant.EPOCH)
        assertThat(foods).filteredOn(FoodForSynchronisation::id, id.id())
            .isNotEmpty
            .allMatch { it.name() == input.name() }
            .allMatch { it.storeUnit() == input.storeUnit() }
            .allMatch { it.toBuy() == input.toBuy() }
    }

    @Test
    fun renameFood() {
        val newFood = FoodAddForm.create(uniqueName, false, Period.ofDays(0), null, unitRepository.anyUnitId.id(), "")
        val id = foodAddService.add(newFood)
        val locationId = locationRepository.createNewLocationType(uniqueName)
        val input =
            FoodForEditing.create(
                id.id(),
                0,
                uniqueName,
                true,
                newFood.expirationOffset().plusDays(1),
                Optional.of(locationId.id()),
                newFood.storeUnit(),
                uniqueName,
            )

        foodEditService.edit(input)

        val foods = updateService.getFood(Instant.EPOCH)
        assertThat(foods).filteredOn(FoodForSynchronisation::id, id.id())
            .filteredOn(FoodForSynchronisation::version, 1)
            .isNotEmpty
            .allMatch { it.name() == input.name() }
            .allMatch { it.storeUnit() == input.storeUnit() }
            .allMatch { it.toBuy() == input.toBuy() }
            .allMatch { it.expirationOffset() == input.expirationOffset() }
            .allMatch { it.description() == input.description() }
    }

    @Test
    fun settingFoodExpirationOffsetToZeroWorks() {
        val newFood = FoodAddForm.create(uniqueName, false, Period.ofDays(5), null, unitRepository.anyUnitId.id(), "")
        val id = foodAddService.add(newFood)
        val locationId = locationRepository.createNewLocationType(uniqueName)
        val input =
            FoodForEditing.create(
                id.id(),
                0,
                uniqueName,
                true,
                Period.ofDays(0),
                Optional.of(locationId.id()),
                newFood.storeUnit(),
                uniqueName,
            )

        foodEditService.edit(input)

        val foods = updateService.getFood(Instant.EPOCH)
        assertThat(foods).filteredOn(FoodForSynchronisation::id, id.id())
            .filteredOn(FoodForSynchronisation::version, 1)
            .isNotEmpty
            .allMatch { it.name() == input.name() }
            .allMatch { it.storeUnit() == input.storeUnit() }
            .allMatch { it.toBuy() == input.toBuy() }
            .allMatch { it.expirationOffset() == input.expirationOffset() }
            .allMatch { it.description() == input.description() }
    }

    @Test
    fun renamingFailsWithWrongVersion() {
        val newFood = FoodAddForm.create(uniqueName, false, Period.ofDays(0), null, unitRepository.anyUnitId.id(), "")
        val newName = uniqueName
        val id = foodAddService.add(newFood)
        val input = FoodForEditing.create(id.id(), 1, newName, true, Period.ZERO, Optional.empty(), 1, uniqueName)

        assertThatExceptionOfType(StatusCodeException::class.java)
            .isThrownBy {
                foodEditService.edit(input)
            }
            .matches { it.statusCode == StatusCode.INVALID_DATA_VERSION }
    }

    @Test
    fun renamingUnknownIdIsReported() {
        val input = FoodForEditing.create(9999, 0, uniqueName, true, Period.ZERO, Optional.empty(), 1, uniqueName)
        assertThatExceptionOfType(StatusCodeException::class.java)
            .isThrownBy {
                foodEditService.edit(input)
            }
            .matches { it.statusCode == StatusCode.NOT_FOUND }
    }

    @Test
    fun deleteFood() {
        val newFood = FoodAddForm.create(uniqueName, false, Period.ofDays(0), null, unitRepository.anyUnitId.id(), "")
        val id = foodAddService.add(newFood)

        foodDeleteService.delete(FoodForDeletion.create(id.id(), 0))

        val foods = updateService.getFood(Instant.EPOCH)
        assertThat(foods).filteredOn(FoodForSynchronisation::id, id.id())
            .isNotEmpty
            .anyMatch { it.transactionTimeEnd().isBefore(Constants.INFINITY) }
    }

    @Test
    fun deletingFailsWithWrongVersion() {
        val newFood = FoodAddForm.create(uniqueName, false, Period.ofDays(0), null, unitRepository.anyUnitId.id(), "")
        val id = foodAddService.add(newFood)
        assertThatExceptionOfType(StatusCodeException::class.java)
            .isThrownBy {
                foodDeleteService.delete(FoodForDeletion.create(id.id(), 9999))
            }
            .matches { it.statusCode == StatusCode.INVALID_DATA_VERSION }
    }

    @Test
    fun deletingUnknownIdIsReported() {
        assertThatExceptionOfType(StatusCodeException::class.java)
            .isThrownBy {
                foodDeleteService.delete(FoodForDeletion.create(9999, 0))
            }
            .matches { it.statusCode == StatusCode.NOT_FOUND }
    }
}
